new routine gchem_insolation replaces duplicate code in dic and bling

mjlosch opened the pull request:
## What changes does this PR introduce?
new routine gchem_insolation replaces duplicate code in dic and bling

## What is the current behaviour? 
- duplicate insolation code in dic and bling
- grid restriction
- see also discussion hidden in PR #876, https://github.com/MITgcm/MITgcm/pull/876#issuecomment-3510316082
0

## What is the new behaviour 
- new routine gchem_insolation replaces duplicate code in dic and bling
- and removes grid restriction

## Does this PR introduce a breaking change? 
No, darwin may want to use this, too.

## Other information:

## Suggested addition to `tag-index`
o pkg/gchem: new routine gchem_insolation.F 
  - replaces duplicate code in pkg/bling and pkg/dic
  - fixes grid restrictions

mjlosch committed:
new routine gchem_insolation replaces duplicate code in dic+bling

jm-c committed:
Merge branch 'master' into gchem_insolation

jm-c committed:
adjust comments

jm-c commented:
Without knowing much about what this "dayfrac" is doing, I think it would be nice if the use of `pkg/cal` (i.e., useCAL=T) was not introducing a discontinuity in `gchem` insolation over time. For instance, towards the end of a 100.yr run that started in 1920, at the last time-step of 2019, `yearSec` will be estimated from a non-leap year (i.e., 2019), but at the next timestep, now in 2020, `yearSec` is estimated from a leap-year.
This will trigger a jump for "dayfrac" from 1 time-step to the next from 25/365 to 291/366.

The other minor issue is just about precision: the astronomical insolation does not know about leap-year or non-leap-year, it just care about the length of the astronomical year (i.e., period of revolution around the sun). And not obvious that the starting time of any simulation is always at a fixed, zero, `dayfrac`.

mjlosch commented:
OK, I guess I was not thinking when I introduced the pkg/cal code. From @jm-c's comment I understand that I should remove this code again.
What about the length of the [sidereal year](https://en.wikipedia.org/wiki/Sidereal_year)? It's 365.256363 days according to Wikipedia. And then there's the mean tropical year of 365.242189 days. In some places we (implicitly) use 365.25 as an average number of days, e.g. in bling_readparms.F, or in a comment in set_defaults.F.
@averdy, @seamanticscience what to you think? What do Paltridge and Platt say (I do not have access to this reference, I assume it is "Paltridge, G. W. and C. M. R. Platt. 1976. Radiative Processes in Meteorology and Climatology. Elsevier Sci. 318 p.")? Or should we just leave it at 360?

mjlosch committed:
remove pkg/cal code and update comments

averdy reviewed (approved):
I tried this in my regional model without issues.

seamanticscience commented:
@mjlosch runtime parameter sounds good. I'm wondering if it would introduce huge inconsistencies if, say `externForcingCycle` is set to 360 days (like tutorial_global_biogeo), while insolation uses 365.25? I think we might get some weird procession if the run is a few decades long?

averdy commented:
@mjlosch @seamanticscience I agree with the runtime parameter idea with default value (365.35 sounds reasonable to me). Could it be set to match externForcingCycle if no runtime parameter is specified? 

It would be great to add code for including the diurnal cycle, and for calculating dayfrac when the start date isn't Jan-01 and pkg/cal is used.

jm-c commented:
@averdy I think you meant 365.25 (and not 365.35) ?
And an alternative to convoluted default settings (since `externForcingCycle` could be used for other things than repeating yearly forcing) would be a weak warning (in STDERR).

jahn commented:
Not sure 365.25 is a good default value.  At least in darwin, this routine is mostly used for simple setups with a 360-day calendar, so no cal package or `TheCalendar='model'`.  I'd prefer 360 days for a simple default, or make it depend on use of the calendar package and the TheCalendar parameter, so 360 unless cal is used and TheCalendar='gregorian', in which case it would be 365.25.

For the AD version of global_oce_biogeo_bling, which has cal on because of ecco but still uses externForcingCycle for all forcing, I would have set TheCalendar to 'model' to make it consistent, then the 360-day default year length for insolation would match externForcingCycle.

mjlosch committed:
change theCalendar to "model", add useCAL=T explicitly

mjlosch committed:
move call cal_set from cal_init_fixed.F to cal_readparms.F

since we may need some of the parameters that are set in cal_set.F
before cal_init_fixed.F is called from initialise_fixed.F.
I could not find any detrimental effect of this change.

mjlosch committed:
add consistency check between calendar and externForcingCycle

- soft warning (no stop) to STDOUT
- only if useEXF=F

mjlosch committed:
new runtime parameter gchem_secondsPerYear

defaults to 360*86400 seconds if useCAL=F
if useCAL=T:
- defaults to 365.25*86400 seconds days if Gregorian calendar is used
- defaults to nDaysNoLeapYear*86400 seconds otherwise

mjlosch committed:
suggestion for a gchem-configation summary (starting point)

jm-c commented:
The move of "CALL CAL_SET" from `cal_init_fixed.F` to `cal_readparms.F` is a bit unfortunate since `endtime` can be adjusted until "set_params.F" is called.

mjlosch commented:
I didn't see that. I need some of the variables set in `cal_set` to generate sensible defaults. Do you have a suggestion of how to solve that? Can we call `cal_set` twice, i.e. in both `cal_readparms.F` and in `cal_init_fixed.F`?

mjlosch committed:
mv cal_set back to cal_init_fixed and adjust gchem_readparms

jm-c commented:
@mjlosch for now, i would prefer solution (1). One day, we might sort out what is really needed and used in `pkg/cal`, split the initialisation in two parts, and then this could be simplified and cleaned up.

jm-c commented on verification/global_oce_biogeo_bling/input_ad/data.pkg:
```diff
@@ -4,7 +4,10 @@
  useECCO        = .TRUE.,
+# this is the default if useECCO=T or usePROFILES = T and ALLOW_CAL is defined
+ useCAL         = .TRUE.,
+# this is the default if ALLOW_CTRL is defined
```
I am not sure about this comment as I cannot find the code where `useGrdchk` is set/reset to True.

mjlosch replied:
You are right, it's the other way around: if `useGrdchk=T`, then set `useCTRL=T`. I'll remove the comment

jm-c reviewed (commented):
The new code related to `gchem_insolation.F` looks good to me. Not sure about the various warning addition, in particular in `cal_summary.F`, since it's not clear to me the relation between `useCAL` and `externForcingCycle` (in fact, the specifications from `pkg/cal` are not relevant for the way `periodicExternalForcing` works).

mjlosch committed:
adjustments following jm-c's suggestions

jm-c commented on pkg/gchem/gchem_check.F:
```diff
@@ -143,6 +158,53 @@ SUBROUTINE GCHEM_CHECK( myThid )
+      IF ( periodicExternalForcing ) THEN
+       IF ( gchem_secondsPeryear .NE. externForcingCycle ) THEN
+        WRITE(msgBuf,'(2(A,F12.1))')
+     &       '** Warning ** GCHEM_INIT_FIXED: gchem_secondsPeryear = ',
```
This printed messgae is from gchem_check.F`, not so much from  GCHEM_INIT_FIXED ; and same for the following.

jm-c reviewed (commented):
I have few comments about `gchem_check.F` addition (see below). 

I also don't know what is the best check for `externForcingCycle` vs `gchem_ForcingCycle`. Maybe, for now, just check if `gchem_secondsPerYear`  is different from `externForcingCycle` ? 
or an alternative would be, if the 2 cycles are equal (i.e., the default) then check if secondsPerYear is equal to `externForcingCycle`, and otherwise (i.e, `gchem_ForcingCycle` is set explicitly) check if secondPerYear is equal to `gchem_ForcingCycle` ?

mjlosch committed:
more fixes and adjustments

jm-c commented:
I took a quick look, it looks good. I am just going to change to print to STDERR instead and after that you could check it that sounds good.

jm-c committed:
Merge branch 'master' into gchem_insolation

jm-c committed:
Merge branch 'master' into gchem_insolation

jm-c committed:
move warnings to STDERR

and adjust printed message

jm-c reviewed (approved):
LGTM

jm-c commented:
@mjlosch Thanks. 
Yesterday I also re-run few testreport (fwd, taf-adm, tap-adm) and the OpenAD test pass the CI tests, so it's good.
I am going to merge this PR later today.

jm-c committed:
document adding gchem_insolaton.F

jm-c commented on pkg/gchem/gchem_check.F:
```diff
@@ -163,16 +232,86 @@ SUBROUTINE GCHEM_CHECK( myThid )
+      gchem_rl = (/gchem_rl1,gchem_rl2,
+     &     gchem_rl3,gchem_rl4,gchem_rl5/)
+      gchem_fileName = (/fileName1,fileName2,
+     &     fileName3,fileName4,fileName5/)
```
Should I disable all `gchem` experiments for this pgf77 test ?
